JSpecifyMode: Optional.ofNullable().map() considers that the values in map are Nullable

Hello ! I'm trying to enable `JSpecifyMode` in my projects. I don't know if it's the same root cause as the previous issus I submitted, #1671 .

For the following method, NullAway reports an error, while the values that are passed to map are always non-null (it's the principle of `Optional`) :
* `extractor.apply(value)` returns a Nullable list
* `Optional.ofNullable()` should produce an `Optional<@NonNull List<String>>` (either with a non null list, or being empty)
* but NullAway (and IntelliJ also without reporting a warning/error) considers that `map()` argument is nullable

<img width="900" height="160" alt="Image" src="https://github.com/user-attachments/assets/1285fc49-25fe-4ce6-ae18-362196907de7" />

```java
  public static <T> Optional<String> convert(
      final T value, final Function<T, @Nullable List<String>> extractor) {
    return Optional.ofNullable(extractor.apply(value)).map(list -> Integer.toString(list.size()));
  }

```

Error :

```
  .../StringUtil.java:31: error: [NullAway] dereferenced expression 'list' is @Nullable
      return Optional.ofNullable(extractor.apply(value)).map(list -> Integer.toString(list.size()));
                                                                                          ^
      (see http://t.uber.com/nullaway )

```

--- a/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
@@ -134,7 +134,8 @@ public final class GenericsChecks {
    * Handles cases where a poly expression (lambda or method reference) is passed as a parameter,
    * and the invoked function has a library model for the corresponding formal. In such cases, we
    * ensure the appropriate formal parameter type from the library model is used when inferring the
-   * type of the poly expression, and cache the result.
+   * type of the poly expression, and cache the result. If inference has already cached a
+   * context-derived type, we preserve it since inference itself applies library models.
    */
   @SuppressWarnings({"ReferenceEquality", "TypeEquals"})
   public void maybeStoreLibraryModeledPolyExpressionType(
@@ -147,6 +148,9 @@ public final class GenericsChecks {
     Preconditions.checkArgument(
         state.getPath().getLeaf() == polyExpressionTree,
         "Expected current path leaf to be the poly expression");
+    if (inferredPolyExpressionTypes.containsKey(polyExpressionTree)) {
+      return;
+    }
     TreePath invocationPath = state.getPath().getParentPath();
     // Skip parentheses around the invocation argument.
     while (invocationPath != null && invocationPath.getLeaf() instanceof ParenthesizedTree) {
@@ -176,10 +180,7 @@ public final class GenericsChecks {
     if (formalParameterType == null || formalParameterType.isRaw()) {
       return;
     }
-    Type polyExpressionType = inferredPolyExpressionTypes.get(polyExpressionTree);
-    if (polyExpressionType == null) {
-      polyExpressionType = ASTHelpers.getType(polyExpressionTree);
-    }
+    Type polyExpressionType = ASTHelpers.getType(polyExpressionTree);
     if (polyExpressionType != null) {
       Type groundTargetType =
           GenericsUtils.groundTargetType(formalParameterType, state, config, handler);

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericLambdaTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericLambdaTests.java
@@ -239,6 +239,30 @@ public class GenericLambdaTests extends NullAwayTestsBase {
         .doTest();
   }
 
+  @Test
+  public void optionalOfNullableRefinesNullableFunctionResult() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            """
+            package com.uber;
+            import java.util.List;
+            import java.util.Optional;
+            import java.util.function.Function;
+            import org.jspecify.annotations.NullMarked;
+            import org.jspecify.annotations.Nullable;
+            @NullMarked
+            class Test {
+              public static <T> Optional<String> convert(
+                  final T value, final Function<T, @Nullable List<String>> extractor) {
+                return Optional.ofNullable(extractor.apply(value))
+                    .map(list -> Integer.toString(list.size()));
+              }
+            }
+            """)
+        .doTest();
+  }
+
   private CompilationTestHelper makeHelper() {
     return makeTestHelperWithArgs(
         JSpecifyJavacConfig.withJSpecifyModeArgs(

